runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java

```diff
@@ -0,0 +1,139 @@
+     *
+     * The future returned does not block explicitly, it aggregates the futures and delegates the blocking.
+     *
+     * In case if some future completes successfully its value will be returned.
+     * In case of termination, the cancel flag will be updated and if any of the futures threw an exception,
+     * ExecutionException will be thrown,  otherwise the future will return null
+     *
+     * @param futures The N futures
+     * @return The composite future
+     */
+    static <R> Future<R> getFirstWinsFuture(CompletableFuture<R>... futures) {
+        return getCompositeFuture(1, futures);
+    }
+
+    private static <R> Future<R> getCompositeFuture(int futuresToWait, CompletableFuture<R>... futures) {
+        return new Future<R>() {
+            private boolean done = false;
+            private boolean canceled = false;
+
+            @Override
+            public R get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+                long until = 0;
+                boolean infinite = (timeout==Long.MAX_VALUE);
+                if (!infinite) {
+                    until = System.nanoTime() + unit.toNanos(timeout);
+                }
+                while (infinite || System.nanoTime() < until) {
+                    Integer lastCompleteFutureIndex = null;
+                    Integer lastExceptionIndex = null;
+                    int numAbnormalExists = 0;
+                    int numCompleteFutures = 0;
+                    CompletableFuture aggregatedFuture = null; // block until some future completes
+                    for (int i = 0; i < futures.length; i++) {
+                        CompletableFuture<R> c = futures[i];
+                        if (!c.isDone()) {
+                            if (aggregatedFuture == null) {
+                                aggregatedFuture = c;
+                            } else {
+                                aggregatedFuture = CompletableFuture.anyOf(aggregatedFuture, c);
+                            }
+                        } else {
+                            if (c.isCancelled() ) {
+                                numAbnormalExists++;
+                            } else if (c.isCompletedExceptionally()) {
+                                numAbnormalExists++;
+                                lastExceptionIndex = i;
+                            } else {
+                                lastCompleteFutureIndex = i;
+                                numCompleteFutures++;
+                            }
+                        }
+                    }
+                    int quorum = futures.length/2+1;
+                    if (numCompleteFutures>=futuresToWait) { // normal exit, quorum
+                        done = true;
+                        return futures[lastCompleteFutureIndex].get();
+                    }
+                    if (numAbnormalExists>=quorum) { // abnormal exit, canceled or failed futures
```

Comment:
If futuresToWait != futures.length/2+1 is it correct to exit when numAbnormalExists>=quorum or should it wait until futuresToWait can no longer be satisfied?